### PR TITLE
Changes to sparse checkout

### PR DESCRIPTION
**Pull request by jim-barber-he**
The existing code checkout downloaded more data than it needed to. This is an attempt to make it more sparse and to fix some potential bugs.

Turn `CHECKOUT_PATHS` into an array so that multiple paths can be used.

Add a step to scan SSH keys of the remote git host.

Add `--filter=blob:none` to the `git clone` so that file contents are not downloaded until needed by git.

If `$BUILDKITE_REPO_MIRROR` is set, then pass it as `--reference` to `git clone`. This is for users using the Buildkite repository mirror functionality.

Add `--depth 1` to the `git fetch` commands for a shallow clone to checkout only necessary data.

Remove the deprecated `git sparse-checkout init` command.

Use `git set` in place of `git add` and pass the `--no-cone` option so that the paths passed in can match files and patterns.

The `git checkout` needs to happen after the `git sparse-checkout` command has been run.
Doing it beforehand checkouts out all files in the repository, which defeats the purpose of performing a sparse checkout.

This is a `bash` script so use `[[ ]]` in place of `[ ]` and there is no need to quote the variables within `[[ ]]` like there was before.

Quote other uses of the variables to fix warnings from shellcheck.

### CODE REVIEW

File: hooks/checkout
Comment (jradtilbrook) [inline hooks/checkout:31]: This makes sense to remove in a way, but its actually still causing the wrong build message in the UI. I think there is some other flow we need to do but I haven't fully researched or tried it out yet.
Basically, without this, the hook is not actually checking out the commit for the build given by `BUILDKITE_COMMIT`. So then below when you fetch, add sparse checkout and then checkout again, all its doing is creating a sparse checkout at the master/main branch because it hasnt switched to the given commit yet.

**Comment by jim-barber-he** [comment]
Potentially this should probably have a plugin option like `no_cone: true` added that defaults to `false` since it isn't a setting recommended by git, but its one I need to be able to match files and wildcards.
Having it as the default of `false` meaning the `--no-cone` option would not be passed to `git clone`.
But I figured I'd worry about that only if other issues around this sparse checkout are solved.

File: hooks/checkout
Comment (jradtilbrook) [reply hooks/checkout:31]: Perhaps using `git switch` is intended for this. Or perhaps even just leave the checkout in place and it won't touch files since its cloned without a checkout. We'll do some testing on our side

Comment (jim-barber-he) [reply hooks/checkout:31]: If this is needed then the final `git checkout` on the last line of this script maybe should become:
```
if [[ ${BUILDKITE_COMMIT} = "HEAD" ]]; then
    git checkout FETCH_HEAD
else
    git checkout "${BUILDKITE_COMMIT}"
fi
```

**Comment by jradtilbrook** [comment]
@jim-barber-he could you close this and reopen to allow us to modify the PR so we can collaborate: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork?

I don't think you can change that setting on an existing PR but if you can then do that

**Comment by jim-barber-he** [comment]
I was unable to find the `Allow edits from maintainers` option both on the existing pull-request, nor when trying to open a new pull request.
My GUI is not matching what that document you linked shows.

**Comment by jradtilbrook** [comment]
Oh because the fork is from an org, not your personal user? Okay kinda makes sense from a security perspective. I'll add it to `main` here and you'll have to pull it into your fork branch.

**Comment by jradtilbrook** [comment]
Alright just merged in #2 if you can update your branch. 
This is the solution that I think works best for the subsequent build without sparse-checkout problem you mentioned in https://forum.buildkite.community/t/agent-3-67-0-has-incorrect-pipeline-headings/3651/21

File: hooks/checkout
Comment (jradtilbrook) [reply hooks/checkout:31]: Yeah I think either way should work. Whether leaving the checkouts in the conditionals here, or reproducing it below after the sparse checkout config

**Comment by jim-barber-he** [comment]
Ok. That change itself won't fix the build heading though. Shall I also make the change I mentioned above? https://github.com/buildkite-plugins/sparse-checkout-buildkite-plugin/pull/1#discussion_r1575482319

**Comment by jradtilbrook** [comment]
> Ok. That change itself won't fix the build heading though. Shall I also make the change I mentioned above? [#1 (comment)](https://github.com/buildkite-plugins/sparse-checkout-buildkite-plugin/pull/1#discussion_r1575482319)

Yeah add that in as well and I think it should be very close!

**Comment by jradtilbrook** [comment]
You can ignore the build failures as well. We can clean them up separately

**Comment by jim-barber-he** [comment]
> Yeah add that in as well and I think it should be very close!

Ok that seemed to fix it.
Tested on this build: https://buildkite.com/healthengineau/sre/builds/264

**Comment by jradtilbrook** [comment]
Awesome. Can you try a rebuild or 2 just to make sure the directory sparseness works with subsequent builds?
And maybe push some empty commits just to make sure the build message does stay correct with subsequent builds as well

**Comment by jim-barber-he** [comment]
I've done a few. It's tricky to make it land on the same agent that it did before though.
Also I push up a change to make the `--no-cone` parameter optional (defaulting to not being used).
And making my initial pipeline look like so:
```
steps:
  - agents:
      queue: build
    command: buildkite-agent pipeline upload
    plugins:
      - HealthengineAU/sparse-checkout:
          no_cone: true
          paths:
            - .buildkite
            - helm*
```
The `helm*` part is just there for testing... I don't need to check that part out for the pipeline upload step.

File: hooks/checkout
Comment (jim-barber-he) [reply hooks/checkout:31]: Can't leave the checkouts here, or else it'll check all files out since the sparse-checkout stuff hasn't been set in motion yet.

**Comment by jradtilbrook** [comment]
Okay nice, could you update the readme with that parameter as I dont quite understand it at the moment. 

Then I think we could merge this in and let you test it out over a few days to see if it does cause any problems and we can follow up with a fix.

We will also follow up with the tests failures and other docs outside of this PR

**Comment by jim-barber-he** [comment]
Ok. Once I do the README change are you happy for me to squash my commits and force-push to clean things up?

**Comment by jradtilbrook** [comment]
> Ok. Once I do the README change are you happy for me to squash my commits and force-push to clean things up?

Yeah certainly 👍

**Comment by jim-barber-he** [comment]
I made a mistake not making a branch for my fork, so I only know how to squash the commits I made after your PR merge adding the environment file.
That's done now with the README update.

**Review by jradtilbrook — APPROVED** [review]

**Comment by jradtilbrook** [comment]
Thats okay. I squash merged it from here anyway. Thanks for this @jim-barber-he I actually think this is a really great plugin for advanced users to improve their build times 👏 
I'm happy we could make it work!!

File: README.md
Comment (jim-barber-he) [inline README.md:19]: When the tests are fixed the `by` in the sentence above should be `be`.

**Pull request merged**
